client/src/main/java/com/tigrisdata/db/client/service/StandardTigrisDBAsyncClient.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2022 Tigris Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tigrisdata.db.client.service;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.tigrisdata.db.api.v1.grpc.Api;
+import com.tigrisdata.db.api.v1.grpc.TigrisDBGrpc;
+import com.tigrisdata.db.client.auth.AuthorizationToken;
+import com.tigrisdata.db.client.config.TigrisDBConfiguration;
+import com.tigrisdata.db.client.error.TigrisDBException;
+import com.tigrisdata.db.client.model.DatabaseOptions;
+import com.tigrisdata.db.client.model.TigrisDBResponse;
+import static com.tigrisdata.db.client.model.TypeConverter.*;
+import com.tigrisdata.db.client.utils.Utilities;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.StatusRuntimeException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class StandardTigrisDBAsyncClient extends AbstractTigrisDBClient
+    implements TigrisDBAsyncClient {
+
+  private final TigrisDBGrpc.TigrisDBFutureStub stub;
+  private final Executor executor;
+
+  private StandardTigrisDBAsyncClient(
+      TigrisDBConfiguration clientConfiguration, AuthorizationToken authorizationToken) {
+    this(clientConfiguration, authorizationToken, Executors.newCachedThreadPool());
+  }
+
+  public StandardTigrisDBAsyncClient(
+      TigrisDBConfiguration clientConfiguration,
+      AuthorizationToken authorizationToken,
+      Executor executor) {
+    super(clientConfiguration, authorizationToken);
+    this.stub = TigrisDBGrpc.newFutureStub(channel);
+    this.executor = executor;
+  }
+
+  @VisibleForTesting
+  StandardTigrisDBAsyncClient(
+      AuthorizationToken authorizationToken,
+      ManagedChannelBuilder<? extends ManagedChannelBuilder> managedChannelBuilder) {
+    super(authorizationToken, managedChannelBuilder);
+    this.stub = TigrisDBGrpc.newFutureStub(channel);
+    this.executor = Executors.newCachedThreadPool();
```

Comment:
does it have any bound on how many threads it will be creating?

Reply:
very good point Himank. So the default here is unbounded however the type of work these threads are going to execute is very light weighted and short lived so I expect this number not to grow too much even in heavy load environment. Still to place the guards - I moved away from lazy cached thread pool to fixed eagerly initialized thread pool.

Reply:
So the main problem would be that if server is slow then we may end up in a situation of lot of threads piling up on the client?